src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+    v |= ~mask; /* Set the unmasked (high) bits. */
+    v = rev(v); /* Reverse. The unmasked bits are now the low bits. */
+    v++;        /* Increment the reversed cursor, flipping the unmasked bits to
+                 * 0 and increments the masked bits. */
+    v = rev(v); /* Reverse the bits back to normal. */
+    return v;
+}
+
+/* The reverse of nextCursor. */
+static size_t prevCursor(size_t v, size_t mask) {
+    v = rev(v);
+    v--;
+    v = rev(v);
+    v = v & mask;
+    return v;
+}
+
+/* Returns 1 if cursor A is less then cursor B, compared in cursor next/prev
+ * order, 0 otherwise. This function can be used to compare bucket indexes in
+ * probing order (since probing order is cursor order) and to check if a bucket
+ * has already been rehashed, since incremental rehashing is also performed in
+ * cursor order. */
+static inline int cursorIsLessThan(size_t a, size_t b) {
+    /* Since cursors are advanced in reversed-bits order, we can just reverse
+     * both numbers to compare them. If a cursor with more bits than the other,
+     * it is not significant, since the more significatnt bits become less
+     * significant when reversing. */
+    return rev(a) < rev(b);
+}
+
+/* Rehashes one bucket. */
+static void rehashStep(hashset *s) {
+    assert(hashsetIsRehashing(s));
+    size_t idx = s->rehashIdx;
+    bucket *b = &s->tables[0][idx];
+    int pos;
+    for (pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+        if (!(b->presence & (1 << pos))) continue; /* empty */
+        void *element = b->elements[pos];
+        uint8_t h2 = b->hashes[pos];
+        /* Insert into table 1. */
+        uint64_t hash;
+        /* When shrinking, it's possible to avoid computing the hash. We can
+         * just use idx has the hash, but only if we know that probing didn't
+         * push this element away from its primary bucket, so only if the
+         * bucket before the current one hasn't ever been full. */
+        if (s->bucketExp[1] < s->bucketExp[0] && !s->tables[0][prevCursor(idx, expToMask(s->bucketExp[0]))].everfull) {
+            hash = idx;
+        } else {
+            hash = hashElement(s, element);
+        }
+        int pos_in_dst_bucket;
+        bucket *dst = findBucketForInsert(s, hash, &pos_in_dst_bucket, NULL);
+        dst->elements[pos_in_dst_bucket] = element;
+        dst->hashes[pos_in_dst_bucket] = h2;
+        dst->presence |= (1 << pos_in_dst_bucket);
+        if (!dst->everfull && bucketIsFull(dst)) {
+            dst->everfull = 1;
+            s->everfulls[1]++;
+        }
+        s->used[0]--;
+        s->used[1]++;
+    }
+    /* Mark the source bucket as empty. */
+    b->presence = 0;
+    /* Bucket done. Advance to the next bucket in probing order, to cover
+     * complete probing chains. Other alternatives are (1) just rehashIdx++ or
+     * (2) in reverse scan order and clear the tombstones while doing so.
+     * (Alternative is to do rehashIdx++.) */
```

Comment:
I guess I'll comment here. Why have probing order be the same as cursor order?

Reply:
Because in scan we need to follow probing chains and not return until we reach the end of a probing chain. This prevents the possibility that we miss an element if rehashing happens between two calls to scan.

Reply:
Maybe that should be documented here then? The comment now doesn't make a lot of sense since the other two options don't work.

For your statement to be true, don't we also need to follow the probing chains until the end during rehashing as well then? Items can be moved to early positions in the chain during rehashing if we are growing the table and items in the current table were placed in buckets later in the chain.

Reply:
Scan needs to guarantee that rehash can't move something between scan calls into scan's already-covered area such that scan will miss it. I did rant at you today in person talking about this, so I won't _rehash_ it here 😜

Reply:
Yeah we can formulate this comment better. I agree the "other alternatives" are not real alternatives if they don't work. Rehash step also doesn't cover complete probe chains in one call. Scan does that though.

This comment needs to explain why we need to do this. I don't remember the reasoning right now.

TBH, I'm not completely sure we do though. When we scan at cursor X, we follow probing chains in both tables to cover all elements that hash to the index that corresponds to cursor X. This includes elements located at X+1 or later (due to probing), and regardless of in which of the two tables it currently is. This seems to imply that the rehashing order doesn't matter.

What am I missing?

@SoftlyRaining you have a test case where you rigged a race between rehashing and scan. Let's add it and see if it breaks?

Reply:
I remember now: we rehash in cursor order to be able to skip already rehashed buckets in scan.

Reply:
I do think that `rehashStep()` needs to follow possible probe chains (Or alternatively `hashsetScan()` needs to check rehashed buckets for `everfull`, but @madolson convinced me that isn't as good.) I'm going to trot out my ascii diagrams again and resume the debate..

I see a bug regarding rehashing and scanning in hashset. Suppose we have an element "Sam" which has been here the whole time - that means scan is guaranteed not to miss Sam. Currently `rehashStep()` does not follow probe chains and can stop between the bucket Sam hashes to and the bucket where Sam is stored due to open addressing. If scan does not access the already-hashed buckets to see the `everfull` flags and continue, it will miss that there's a probe chain and Sam has been moved forward, so it may also stop before reaching Sam but after the bucket that Sam hashes to.

```
      Sam (☺) hashes here in old table
      |
old:  E E ☺ .
new:  .......
        |
        RehashIdx
        |
        cursor
```
Then between `hashsetScan()` calls `rehashStep()` moves Sam directly to the bucket he hashes to in the new table, in a location that scan believes it has already covered. Thus scan misses Sam, and the guarantee is broken:
```
       Sam (☺) hashes here in new table
       |
old:  E E . .
new:  .☺.....
        |   |
        |   RehashIdx
        |
        cursor
```

If we know there's no probe chains where we're at, we also know that no elements can move to an earlier bucket than our current position due to rehashing. If rehashing follows probe chains, then any element that hashes to an already-rehashed bucket must already have been rehashed, making it safe to scan to skip already-rehashed buckets.

Reply:
> Currently `rehashStep()` does not follow probe chains and can stop between the bucket Sam hashes to and the bucket where Sam is stored due to open addressing. If scan does not access the already-hashed buckets to see the `everfull` flags and continue, it will miss that there's a probe chain and Sam has been moved forward, so it may also stop before reaching Sam but after the bucket that Sam hashes to.
> 
> ```
>       Sam (☺) hashes here in old table
>       |
> old:  E E ☺ .
> new:  .......
>         |
>         RehashIdx
>         |
>         cursor
> ```

I think you missed this: `rehashStep` doesn't clear the `everfull` flags when moving elements from the old to the new table.

If the buckets marked with E have been rehashed, they do not contain any elements, but they still have the `everfull` flag, so scan won't stop between these buckets. Scan will reach Sam on this probe chain.

Reply:
I did see that `everfull` flags are not cleared by rehash. I agree that scan would need to iterate through already-scanned buckets to check the `everfull` flag, not skip them.

Here's the line I think you missed: Currently in hashsetScan() at hashset.c:1302 we have the line below. When it is growing the table and the bucket has already been rehashed, it is not checking `everfull` flag, so scan will (incorrectly) stop between these buckets.
```
if (table_small == 0 && !cursorIsLessThan(cursor, s->rehash_idx)) {
```

A quick fix is to just remove the if statement. A better fix is to make `rehashStep()` not stop in the middle of possible probe chains. If `rehashStep()` follows probe chains it guarantees that if you are behind `rehashIdx` any probe chain you might see has already been rehashed to the new table, allowing us to skip any rehashed bucket. (And the rehashing is work we have to do anyway.)

Reply:
https://github.com/valkey-io/valkey/pull/1217

Reply:
I'm convinced now. Thanks!

Reply:
Chainged to chaining. This problem disappeared.